File: brain/midonet-brain/src/main/java/org/midonet/brain/services/vxgw/VxLanGatewayService.java
/*
 * Copyright 2014 Midokura SARL
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.midonet.brain.services.vxgw;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Random;
import java.util.Set;
import java.util.UUID;

import javax.annotation.Nonnull;

import com.google.common.base.Predicate;
import com.google.common.collect.Collections2;
import com.google.inject.Inject;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import rx.Subscription;
import rx.functions.Action1;

import org.midonet.brain.ClusterMinion;
import org.midonet.brain.ClusterNode;
import org.midonet.brain.MinionConfig;
import org.midonet.brain.services.vxgw.monitor.BridgeMonitor;
import org.midonet.brain.services.vxgw.monitor.DeviceMonitor;
import org.midonet.brain.southbound.vtep.VtepConstants;
import org.midonet.brain.southbound.vtep.VtepDataClientFactory;
import org.midonet.brain.southbound.vtep.VtepNotConnectedException;
import org.midonet.brain.southbound.vtep.VtepStateException;
import org.midonet.cluster.DataClient;
import org.midonet.cluster.EntityIdSetEvent;
import org.midonet.cluster.data.Bridge;
import org.midonet.cluster.data.VTEP;
import org.midonet.cluster.data.VtepBinding;
import org.midonet.cluster.data.ports.VxLanPort;
import org.midonet.config.ConfigBool;
import org.midonet.config.ConfigGroup;
import org.midonet.config.ConfigString;
import org.midonet.midolman.serialization.SerializationException;
import org.midonet.midolman.state.NoStatePathException;
import org.midonet.midolman.state.StateAccessException;
import org.midonet.midolman.state.ZookeeperConnectionWatcher;
import org.midonet.packets.IPv4Addr;

/**
 * A service to integrate a Midonet cloud with hardware VTEPs.
 */
public class VxLanGatewayService extends ClusterMinion {

    private static final Logger log =
        LoggerFactory.getLogger(VxLanGatewayService.class);

    // Client for Midonet configuration store
    private final DataClient midoClient;

    // Zookeeper connection watcher
    private ZookeeperConnectionWatcher zkConnWatcher;

    // Provides vtep clients
    private final VtepDataClientFactory vtepDataClientFactory;

    // Index of VxlanGwBrokers for each VTEP
    private final Map<IPv4Addr, VxLanGwBroker> vxlanGwBrokers = new HashMap<>();

    // Monitors
    private HostStatePublisher hostMonitor = null;
    private TunnelZoneStatePublisher tunnelZoneMonitor = null;
    private VtepStatePublisher vtepMonitor = null;
    private BridgeMonitor bridgeMonitor = null;

    // Subscription observables
    private List<Subscription> subscriptions = new ArrayList<>();

    // Random number generator.
    private final Random random;

    // Cluster Node context
    private final ClusterNode.Context nodeCtx;

    @Inject
    public VxLanGatewayService(
        @Nonnull ClusterNode.Context nodeCtx,
        @Nonnull DataClient midoClient,
        @Nonnull VtepDataClientFactory vtepDataClientFactory,
        @Nonnull ZookeeperConnectionWatcher zkConnWatcher) {
        this(nodeCtx, midoClient, vtepDataClientFactory, zkConnWatcher, new Random());
    }

    public VxLanGatewayService(
        @Nonnull ClusterNode.Context nodeCtx,
        @Nonnull DataClient midoClient,
        @Nonnull VtepDataClientFactory vtepDataClientFactory,
        @Nonnull ZookeeperConnectionWatcher zkConnWatcher,
        @Nonnull Random random) {
        super(nodeCtx); // Scala-induced awkwardness
        this.nodeCtx = nodeCtx;
        this.midoClient = midoClient;
        this.zkConnWatcher = zkConnWatcher;
        this.vtepDataClientFactory = vtepDataClientFactory;
        this.random = random;
    }

    @Override
    protected void doStart() {
        log.info("Starting up the VXLAN gateway service");

        // Set-up monitoring
        try {
            hostMonitor = new HostStatePublisher(midoClient, zkConnWatcher);
            tunnelZoneMonitor = new TunnelZoneStatePublisher(midoClient,
                                                             zkConnWatcher,
                                                             hostMonitor,
                                                             random);
            vtepMonitor = new VtepStatePublisher(midoClient, zkConnWatcher,
                                                 nodeCtx.nodeId());
            bridgeMonitor = new BridgeMonitor(midoClient, zkConnWatcher);
        } catch (DeviceMonitor.DeviceMonitorException e) {
            log.error("Service failed: cannot set up the device monitors.", e);
            shutdown();
            notifyFailed(e);
            return;
        }

        // VTEP subscriptions
        subscriptions.add(vtepMonitor.getAcquireObservable().subscribe(
            new Action1<VtepState>() {
                @Override
                public void call(VtepState vtep) {
                    onVtepAcquired(vtep.vtepIp);
                }
            }));
        subscriptions.add(vtepMonitor.getReleaseObservable().subscribe(
            new Action1<VtepState>() {
                @Override
                public void call(VtepState vtep) {
                    onVtepReleased(vtep.vtepIp);
                }
            }));


        // Bridge subscriptions
        subscriptions.add(bridgeMonitor.getEntityIdSetObservable().subscribe(
            new Action1<EntityIdSetEvent<UUID>>() {
                @Override
                public void call(EntityIdSetEvent<UUID> event) {
                    switch (event.type) {
                        case CREATE:
                        case STATE:
                            onBridgeCreatedOrUpdated(null, event.value);
                            break;
                    }
                }
            }));
        subscriptions.add(bridgeMonitor.getEntityObservable().subscribe(
            new Action1<Bridge>() {
                @Override
                public void call(Bridge bridge) {
                    onBridgeCreatedOrUpdated(bridge, bridge.getId());
                }
            }));

        // Notify the entities initial state: first VTEPs and then bridges.
        vtepMonitor.notifyState();
        bridgeMonitor.notifyState();

        log.info("Service started");
        notifyStarted();
        this.awaitRunning();
    }

    @Override
    protected void doStop() {
        log.info("Service stopped");
        shutdown();
        notifyStopped();
    }

    /* Cleanup service state */
    private synchronized void shutdown() {
        // Dispose the monitor: must call before un-subscribing in order to
        // receive notifications for any clean-up.
        hostMonitor.dispose();
        tunnelZoneMonitor.dispose();
        vtepMonitor.dispose();

        // Un-subscribe.
        for (Subscription subscription : subscriptions) {
            if (!subscription.isUnsubscribed()) {
                subscription.unsubscribe();
            }
        }
        subscriptions.clear();

        hostMonitor = null;
        tunnelZoneMonitor = null;
        vtepMonitor = null;
        bridgeMonitor = null;

        // All gateway brokers should have been cleaned up when disposing the
        // VTEP monitor.
        assert(vxlanGwBrokers.isEmpty());
    }

    /**
     * Safe method called when the VXGW service acquires the ownership of a
     * VTEP.
     * @param vtepIp The VTEP management address.
     */
    private void onVtepAcquired(final IPv4Addr vtepIp) {
        try {
            onVtepCreatedUnsafe(midoClient.vtepGet(vtepIp));
        } catch (NoStatePathException e) {
            log.warn("VTEP {} does not exist in storage (ignoring)", vtepIp, e);
        } catch (StateAccessException e) {
            log.warn("Cannot retrieve the VTEP state {} from storage (retrying)",
                     vtepIp, e);
            zkConnWatcher.handleError(
                String.format("VxLanGatewayService %s pairing VTEP %s",
                              nodeCtx.nodeId(), vtepIp),
                new Runnable() {
                    @Override
                    public void run() {
                        onVtepAcquired(vtepIp);
                    }
                },
                e);
        } catch (SerializationException e) {
            log.error("Failed to deserialize resource VTEP state {} (aborting)",
                      vtepIp, e);
        } catch (VtepStateException e) {
            log.error("Failed to connect to VTEP {} because of inconsistent"
                      + "connection state (aborting)", e.vtep, e);
        }
    }

    /**
     * Unsafe method called when the VXGW service acquires the ownership of a
     * VTEP. The method creates the VXLAN gateway broker corresponding to a
     * VTEP, and pairs it with a MidoNet VXLAN peer.
     * @param vtep The VTEP.
     * @return The VXLAN gateway broker corresponding to this VTEP.
     */
    private synchronized VxLanGwBroker onVtepCreatedUnsafe(VTEP vtep)
        throws StateAccessException, SerializationException,
               VtepStateException {

        IPv4Addr mgmtIp = vtep.getId();
        log.debug("Starting VXLAN gateway pairing for VTEP {}", mgmtIp);

        // If a gateway broker already exists, do nothing.
        if (vxlanGwBrokers.containsKey(mgmtIp)) {
            log.debug("VXLAN gateway broker is already set up for VTEP {}",
                      mgmtIp);
            return null;
        }

        // Get the tunnel zone corresponding to this VTEP.
        TunnelZoneState tunnelZone =
            tunnelZoneMonitor.getOrTryCreate(vtep.getTunnelZoneId());
        if (null == tunnelZone) {
            // If cannot create the tunnel zone state for
            log.error("Cannot create tunnel zone {} state for VTEP {}. The "
                      + "VXLAN service will initialize the VTEP without a "
                      + "flooding proxy.",
                      vtep.getTunnelZoneId(), mgmtIp);
        }

        // Initialize the VTEP and wire peers.
        VxLanGwBroker vxGwBroker = new VxLanGwBroker(
            midoClient, vtepDataClientFactory, vtep,
            tunnelZone, UUID.randomUUID());
        vxlanGwBrokers.put(mgmtIp, vxGwBroker);

        return vxGwBroker;
    }

    /**
     * Handles the release of ownership of a VTEP, by terminating the gateway
     * broker associated to the VTEP.
     * @param vtepIp The VTEP management address.
     */
    private void onVtepReleased(IPv4Addr vtepIp) {
        log.info("Release VTEP {} by {}", vtepIp, nodeCtx.nodeId());
        VxLanGwBroker broker = vxlanGwBrokers.remove(vtepIp);
        if (broker != null) {
            try {
                broker.terminate();
            } catch (VtepStateException e) {
                log.warn("Broker for VTEP {} failed to disconnect gracefully.",
                         e.vtep);
            }
        }
    }

    /**
     * Safe method that handles the update of an existing bridge. Updates may
     * include the creation and deletion of bridge VXLAN ports.
     * @param bridge The bridge.
     * @param id The bridge identifier.
     */
    private void onBridgeCreatedOrUpdated(final Bridge bridge, final UUID id) {
        log.debug("Bridge created or updated: {}", id);

        try {
            onBridgeUpdatedUnsafe(
                null == bridge ? midoClient.bridgesGet(id) : bridge);
        } catch (NoStatePathException e) {
            log.warn("Bridge {} does not exist in storage (ignoring)",
                     id, e);
        } catch (StateAccessException e) {
            log.warn("Cannot retrieve bridge state {} from storage (retrying)",
                     id, e);
            zkConnWatcher.handleError(
                String.format("VxLanGatewayService %s assigning bridge %s",
                              nodeCtx.nodeId(), id),
                new Runnable() {
                    @Override
                    public void run() {
                        onBridgeCreatedOrUpdated(bridge, id);
                    }
                }, e);
        } catch (VxLanPeerConsolidationException e) {
            log.error("Failed to consolidate VTEP state {} (aborting)",
                      id, e);
        } catch (SerializationException e) {
            log.error("Failed to deserialize resource state {} (aborting)",
                      id, e);
        } catch (VtepNotConnectedException e) {
            log.error("Failed to consolidate VTEP {} state for bridge {} "
                      + "because VTEP is not connected", e.vtep, id);
        }
    }

    /**
     * Handles the creation and update of a bridge, by assigning the bridge
     * to the corresponding MidoNet VXLAN peer.
     * @param bridge The bridge.
     */
    private void onBridgeUpdatedUnsafe(Bridge bridge)
        throws StateAccessException, SerializationException,
               VxLanPeerConsolidationException, VtepNotConnectedException {

        final UUID id = bridge.getId();

        // Get the VXLAN port identifier for this bridge.
        List<UUID> vxlanPortIds = bridge.getVxLanPortIds();
        if (vxlanPortIds == null || vxlanPortIds.isEmpty()) {
            log.debug("Bridge {} does not any port towards VTEP, ignoring", id);
            return;
        }

        UUID vxLanPortId = vxlanPortIds.get(0);
        VxLanPort vxLanPort = (VxLanPort)midoClient.portsGet(vxLanPortId);
        if (null == vxLanPort) {
            log.warn("Cannot retrieve VXLAN port for port ID {} (aborting)",
                     vxLanPortId);
            return;
        }

        IPv4Addr vtepIp = vxLanPort.getMgmtIpAddr();
        VxLanGwBroker broker = vxlanGwBrokers.get(vtepIp);
        if (null == broker) {
            log.info("Unknown or unowned VXLAN broker for VTEP {} (aborting)",
                     vtepIp);
            return;
        }

        if (!broker.midoPeer.knowsBridgeId(id)) {
            log.info("Consolidating VTEP configuration for new bridge "
                     + "{}", id);
            String lsName = VtepConstants.bridgeIdToLogicalSwitchName(id);
            int vni = vxLanPort.getVni();
            org.opendaylight.ovsdb.lib.notation.UUID lsUuid =
                broker.vtepPeer.ensureLogicalSwitchExists(lsName, vni);

            broker.vtepPeer.renewBindings(
                lsUuid,
                Collections2.filter(
                    midoClient.vtepGetBindings(vtepIp),
                    new Predicate<VtepBinding>() {
                        @Override
                        public boolean apply(VtepBinding b) {
                            return b != null && id.equals(b.getNetworkId());
                        }
                    }
                )
            );
        }

        log.info("Monitoring bridge {} for mac updates", id);
        if (broker.midoPeer.watch(id)) {
            broker.vtepPeer.advertiseMacs();
            broker.midoPeer.advertiseFloodingProxy(id);
        }
    }

    /**
     * Returns the host identifier that is designated as the current flooding
     * proxy for the specified tunnel zone (used for testing).
     * @param tzoneId The tunnel zone identifier.
     * @return The host identifier, or null if the tunnel zone does not exist,
     * or the tunnel zone does not have a flooding proxy.
     */
    protected UUID getFloodingProxy(UUID tzoneId) {
        TunnelZoneState tunnelZone = tunnelZoneMonitor.get(tzoneId);
        return (null != tunnelZone && tunnelZone.hasFloodingProxy()) ?
               tunnelZone.getFloodingProxy().id : null;
    }

    /**
     * Gets the set of currently owned VTEPs (used for testing).
     * @return The set of VTEP IP addresses.
     */
    protected Set<IPv4Addr> getOwnedVteps() {
        return vxlanGwBrokers.keySet();
    }

    /** Configuration for the VxGWService Minion. */
    @ConfigGroup(VxGWServiceConfig.configGroup)
    public interface VxGWServiceConfig extends MinionConfig<VxLanGatewayService> {

        public static final String configGroup = "vxgw";

        @ConfigBool(key = "enabled", defaultValue = false)
        public boolean isEnabled();

        @ConfigString(key = "with")
        public String minionClass();
    }
}

File: cluster/midonet-cluster/src/main/scala/org/midonet/cluster/services/vxgw/VxlanGatewayService.scala
/*
 * Copyright 2015 Midokura SARL
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.midonet.cluster.services.vxgw

import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.Executors.newSingleThreadExecutor
import java.util.concurrent.TimeUnit.SECONDS
import java.util.{Random, UUID}

import scala.collection.JavaConversions._
import scala.util.{Failure, Success, Try}

import com.codahale.metrics.MetricRegistry
import com.google.inject.Inject
import org.apache.curator.framework.CuratorFramework
import org.apache.curator.framework.recipes.leader.{LeaderLatch, LeaderLatchListener}
import org.slf4j.LoggerFactory
import rx.schedulers.Schedulers
import rx.{Observer, Subscription}

import org.midonet.cluster.EntityIdSetEvent.Type._
import org.midonet.cluster.southbound.vtep.VtepDataClientFactory
import org.midonet.cluster.{ClusterMinion, ClusterNode, DataClient, EntityIdSetEvent, _}
import org.midonet.midolman.state.Directory.DefaultTypedWatcher
import org.midonet.midolman.state.{StateAccessException, ZookeeperConnectionWatcher}
import org.midonet.util.concurrent.NamedThreadFactory
import org.midonet.util.functors._

/** An implementation of the VxLAN Gateway Service that supports high
  * availability on the hardware VTEPs. The service allows making bindings
  * from ports/vlan pairs in different hardware VTEPs forming a single Logical
  * Switch, and synces Mac-Port pairs across both MidoNet and all the VTEPs that
  * participate in the Logical Switch.
  *
  * Instances of the service coordinate in an active-passive configuration using
  * Zookeeper. Only one node will be elected as leader and perform all VxLAN
  * Gateway management. When a node loses leadership (voluntarily, or because
  * of a failure, partition, etc.) a different instance will be elected and
  * take over the management. */
class VxlanGatewayService @Inject()(nodeCtx: ClusterNode.Context,
                                    dataClient: DataClient,
                                    zkConnWatcher: ZookeeperConnectionWatcher,
                                    vtepDataClientFactory: VtepDataClientFactory,
                                    curator: CuratorFramework,
                                    metrics: MetricRegistry,
                                    conf: ClusterConfig)
extends ClusterMinion(nodeCtx) {

    private val log = LoggerFactory.getLogger(vxgwLog)

    // TODO: take these out to a service metrics container
    private val networkCount = metrics.counter(s"${conf.vxgw.PREFIX}.networks")
    private val vxgwCount = metrics.counter(s"${conf.vxgw.PREFIX}.vxgws")
    def numNetworks: Long = networkCount.getCount
    def numVxGWs: Long = vxgwCount.getCount

    // Executor on which we schedule tasks to release the ZK event thread.
    private val executor = newSingleThreadExecutor(
        new NamedThreadFactory("vxgw-gateway-initializer")
    )

    // Latch to coordinate active-passive with other instances
    private val LEADER_LATCH_PATH = "/midonet/vxgw/leader-latch"
    private val leaderLatch = new LeaderLatch(curator, LEADER_LATCH_PATH,
                                              nodeCtx.nodeId.toString)

    // Index of VxLAN Gateway managers for each neutron network with bindings
    // to hardware VTEPs.
    private val managers = new ConcurrentHashMap[UUID, VxlanGatewayManager]()

    // Watch bridge creations and deletions.
    private val networkUpdateMonitor =
        new ConcurrentHashMap[UUID, DefaultTypedWatcher]()

    // A monitor that notifies on network creation, which we will examine and
    // decide whether we need a new VxLAN Gateway manager.
    private var networkSub: Subscription = _

    // VTEP controllers, lazy initialized when starting the service, not before.
    // No need to make them volatile as the three points where they are used
    // after assigning are single-threaded.
    private var hostsState: HostStatePublisher = _
    private var tzState: TunnelZoneStatePublisher = _
    private var vteps: VtepPool = _

    // An observer that bootstraps a new VxLAN Gateway service whenever a
    // neutron network that has bindings to hardware VTEP(s) is created.
    // Note that updates are handled by the `networkWatcher` placed on each
    // individual network.
    private val bridgesWatcher = new Observer[EntityIdSetEvent[UUID]] {
        override def onCompleted(): Unit = {
            log.warn("Unexpected: networks watcher completed, this indicates " +
                     "that all networks were deleted!")
        }
        override def onError(e: Throwable): Unit = {
            log.warn("Error on network stream, retry subscription..", e)
            executor submit monitorReset
        }
        override def onNext(t: EntityIdSetEvent[UUID]): Unit = {
            if (t.`type` != DELETE) {
                bootstrapIfInVxlanGateway(t.value) // creation, or startup
            }
            // Deletes are dealt with by each individual network watcher
        }
    }

    // Monitorization of networks in the system.  This will load the existing
    // ones at startup, and keep watching while alive for updates.
    private val bridgeBufSize = conf.vxgw.networkBufferSize

    private val alertOverflow = makeAction0 {
        val ex = new IllegalStateException(
            "Network monitoring failed due to buffer overflow. " +
            "Please raise cluster.vxgw.network_buffer_size -  this value" +
            "should be an order of magnitude higher than the expected "+
            "number of virtual networks in your system.  The service will " +
            "require a restart.  Buffer size is now " + bridgeBufSize)
        log.error("Terminating service: ", ex)
        notifyFailed(ex)
        shutdown()
    }

    private val monitorReset: Runnable = makeRunnable {
        log.info("Watching for new VxLAN Gateways")
        val monitor = dataClient.bridgesGetUuidSetMonitor(zkConnWatcher)
        networkSub = monitor.getObservable
                            .onBackpressureBuffer(bridgeBufSize, alertOverflow)
                            .observeOn(Schedulers.from(executor))
                            .subscribe(bridgesWatcher)
        monitor.notifyState()
    }


    /** Create a watcher for changes on the given network that will detect
      * when it becomes part of a VxLAN Gateway and bootstrap the management
      * process. */
    private def newNetworkWatcher(id: UUID) = new DefaultTypedWatcher {
        override def pathDataChanged(path: String): Unit = {
            executor submit makeRunnable { bootstrapIfInVxlanGateway(id) }
        }
        override def pathDeleted(path: String): Unit = networkDeleted(id)
    }

    /** Load the given network and bootstrap the VxGW Manager if it's now
      * part of a VxLAN Gateway.  Otherwise just keep watching updates on the
      * network just in case new bindings to a VTEP are added. */
    private def bootstrapIfInVxlanGateway(id: UUID): Unit = {
        Try {
            // Load the bridge, setting up a watcher if it's the first time seen
            val newWatcher = newNetworkWatcher(id)
            val prevWatcher = networkUpdateMonitor.putIfAbsent(id, newWatcher)
            if (prevWatcher == null) {
                log.info(s"Network created $id")
                networkCount.inc()
                dataClient.bridgeGetAndWatch(id, newWatcher)
            } else {
                dataClient.bridgeGetAndWatch(id, prevWatcher)
            }
        } match {
            case Success(b) if b.getVxLanPortIds == null ||
                               b.getVxLanPortIds.isEmpty =>
                log.info("VxGW is no more")
                if (log.isTraceEnabled) {
                    log.trace(s"Network ${b.getId} changed but isn't bound " +
                              s" to any VTEPs. Watching $numNetworks networks" +
                              s" and $numVxGWs VxLAN Gateways")
                }
            case Success(b) =>
                initVxlanGatewayManager(b.getId)
            case Failure(e: StateAccessException) =>
                zkConnWatcher.handleError("Retry create VxLAN Gateway manager" +
                                          s" for network $id",
                                          makeRunnable {
                                              bootstrapIfInVxlanGateway(id)
                                          }, e)
            case Failure(e) =>
                log.error(s"Error starting VxGW monitor for network $id", e)
        }
    }

    /** React to the deletion of a network by cleaning any existing watcher as
      * well as the VxLAN Gateway manager, should one exist. */
    private def networkDeleted(id: UUID): Unit = {
        log.info(s"Network $id deleted, now watching $numNetworks")
        networkCount.dec()
        networkUpdateMonitor.remove(id)
        // If the network is part of a vxgw, the manager will take care of the
        // cleanup.
    }

    /** Create and run a new VxlanGatewayManager for the given network id */
    private def initVxlanGatewayManager(id: UUID) {
        val nw = new VxlanGatewayManager(id, dataClient, vteps,
                                         tzState, zkConnWatcher,
                                         () => {
                                             managers.remove(id)
                                             vxgwCount.dec()
                                         })
        if (managers.putIfAbsent(id, nw) == null) {
            log.info(s"Network $id is a new VxLAN Gateway (total: $numVxGWs)")
            vxgwCount.inc()
            nw.start()
        } else {
            log.debug(s"VxLAN Gateway manager already exists for network $id")
        }

    }

    // A callback when a change in this node's leadership status happens
    private val latchListener = new LeaderLatchListener {
        override def isLeader(): Unit = {
            log.info("I am the VxLAN Gateway leader! \\o/")
            executor submit monitorReset
        }
        override def notLeader(): Unit = {
            log.info("I am no longer VxLAN Gateway leader, going passive")
            becomePassive()
        }
    }

    override def doStart(): Unit = {
        log.info("Starting service")

        // Watch hosts and tunnel zones in order to discover flooding proxies
        hostsState = new HostStatePublisher(dataClient, zkConnWatcher)
        tzState = new TunnelZoneStatePublisher(dataClient, zkConnWatcher,
                                               hostsState, new Random())
        vteps = new VtepPool(nodeCtx.nodeId, dataClient, zkConnWatcher,
                             tzState, vtepDataClientFactory)

        leaderLatch.addListener(latchListener)
        leaderLatch.start()
        notifyStarted()
    }

    /** Makes the node become passive but not stop. */
    private def becomePassive(): Unit = {
        log.info("Node suspends VxLAN Gateway management")
        if (networkSub != null) {
            networkSub.unsubscribe()
        }
        managers.values().foreach { _.terminate() }
        if (hostsState != null) {
            hostsState.dispose()
        }
        if (tzState != null) {
            tzState.dispose()
        }
    }

    override def doStop(): Unit = {
        try {
            shutdown()
            notifyStopped()
        } catch {
            case t: Throwable =>
                log.error("Failed to shutdown executor", t)
                Thread.currentThread().interrupt()
                notifyFailed(t)
        }
    }

    private def shutdown(): Unit = {
        becomePassive()
        executor.shutdown()
        if (!executor.awaitTermination(5, SECONDS)) {
            log.warn("Failed to stop network monitor orderly, insisting..")
            executor.shutdownNow()
        }
        leaderLatch.removeListener(latchListener)
        leaderLatch.close()
    }

}

